src/liballoc/collections/btree/map.rs

```diff
@@ -207,6 +207,60 @@ impl<K: Clone, V: Clone> Clone for BTreeMap<K, V> {
             clone_subtree(self.root.as_ref())
         }
     }
+
+    fn clone_from(&mut self, other: &Self) {
+        BTreeClone::clone_from(self, other);
+    }
+}
+
+trait BTreeClone {
+    fn clone_from(&mut self, other: &Self);
+}
+
+impl<K: Clone, V: Clone> BTreeClone for BTreeMap<K, V> {
+    default fn clone_from(&mut self, other: &Self) {
+        *self = other.clone();
+    }
+}
+
+impl<K: Clone + Ord, V: Clone> BTreeClone for BTreeMap<K, V> {
+    fn clone_from(&mut self, other: &Self) {
+        // This truncates `self` to `other.len()` by calling `split_off` on
+        // the first key after `other.len()` elements if it exists
+        if let Some(key) = {
+            if self.len() > other.len() {
+                let diff = self.len() - other.len();
+                if diff <= other.len() {
+                    self.iter().nth_back(diff - 1).map(|pair| (*pair.0).clone())
+                } else {
+                    self.iter().nth(other.len()).map(|pair| (*pair.0).clone())
+                }
+            } else {
+                None
+            }
+        } {
+            self.split_off(&key);
```

Comment:
I feel that a lot of duplicate work being done here. We first find the nth key, the pass it to `split_off` so that it can find the position of that key (again!), after which it actually splits the tree. Would it be possible to avoid doing the key lookup twice?

Reply:
Not that I can see, unfortunately! However, `split_off` doesn't exactly find the key and then split the tree. Because `split_off` has a target key, it can split the tree as it's searching for the key position on the edges it descends on. In theory it would be better to have a `split_after` function that splits off the first n keys, but in practice the `BTree` doesn't have any subtree size information available (and traverses the whole tree every time it needs to recalculate length as a result).  Similarly, it would definitely be more efficient if `nth` was specialized for `BTree` iterators, but I don't see an easy way to do that. Please let me know if you have any ideas about how to avoid iterating through (up to) half the tree!

Reply:
My point is that you are traversing the tree twice: once in the `nth` call in the iterator, and once in `split_off`. However fixing this is likely to be very complicated, so it's fine to leave it as it is for now.